### PR TITLE
Make MoreUnit easier accessible for contributors

### PR DESCRIPTION
**Pull request by rherrmann**
The change adds target platforms for Eclipse versions 4.3 - 4.6.
The master pom uses the eclipse-4.3 (Kepler) target platform so that
the behavior of the tycho build remains unchanged.
Changing MoreUnitAnnotationModel became necessary because of compile
errors with the 4.6 target platform.

### CODE REVIEW

**Comment by apupier** [comment]
Can you elaborate on Compile error for MoreUnitAnnotationModel? The current implementation seems strange for me as we are making unchecked Cast

**Comment by rherrmann** [comment]
In 4.6 the return type of `IAnnotationModel::getAnnotationIterator` has changed
from 
```java
Iterator getAnnotationIterator();
```
to 
```java
Iterator<Annotation> getAnnotationIterator()
```

The `annotations`  list was defined as `List<Annotation> annotations` and therefore a compile error was issued:
```text
Type mismatch: cannot convert from Iterator<Annotation> to Iterator<MoreUnitAnnotation>
``` 

From what I understand the `MoreUnitAnnotationModel` was never meant to hold anything else than `MoreUnitAnnotation`s. There is only one place in `annotateTestedMethods` where annotations are added. Therefore I think the casts are safe.

My primary goal was to fix #6 thus this fix may seem a bit hasty. I'm open for suggestions on how to improve the situation.

**Comment by apupier** [comment]
I better understand the issue now.
The best would be to modify the IAnnotationModel in Eclipse itself, asking for this kind of interface Iterator<? extends Annotation> getAnnotationIterator();
As it is still in 4.6 it might be possible but seems not trivial as as after a first try it seems that we need to change it in several places.

Another solution is to create a List of the supertype in the method such as:
    public Iterator<Annotation> getAnnotationIterator()
    {
        return new ArrayList<Annotation>(annotations).iterator();
    }
It creates another list for not a lot but it allows to keep the specific type everywhere else. Seems to provide a better readability from my POV.

What do you think?

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+ 
+3. The necessary projects are located at the root of the repository. 
+Import all of them into your workspace.
+
+4. The `org.moreunit.build` project contains suitable [Target Platform Definitions](http://help.eclipse.org/juno/index.jsp?topic=%2Forg.eclipse.pde.doc.user%2Fconcepts%2Ftarget.htm).
+Open one of them and select the _Set as Target Platform_ link.
+
+
+Now you are ready to hack the sources.
+If you whish to verify the setup, you can run the test suite. The launch configuration to run all tests is called _All Tests_
```
Comment (apupier) [inline CONTRIBUTING.md]: I'm not very familiar with this project but I think that there are 2 test suites and they are called _MoreUnit Tests_ and _MoreUnit Mock Support Integration Tests_

Comment (rherrmann) [inline CONTRIBUTING.md]: Sorry, this sentence is plain wrong - I've taken it from another project of mine.
As far as I can see, MoreUnit currently has no launch configuration to run all tests (include the SWTBot tests). I would rewrite the phrase as:
> If you wish to verify the setup, you can run the Maven/Tycho build with the _mvn clean install_ launch configuration.

**Comment by rherrmann** [comment]
Excellent idea to return a new list. Though it requires a `@SuppressWarnings({ "rawtypes", "unchecked" })`, the change is much more local, plus it is the right step towards fixing issue #10 

While I agree that this actually should be fixed in the platform, such change will probably not be available short term. 

I will upload the respective change and also a changed `CONTRIBUTION.md`..

**Comment by apupier** [comment]
+1 for merge

**Comment by ndemengel** [comment]
None of the targets seems to work for me, running "mvn clean install" gives me errors such as "org.eclipse.tycho.p2.target.facade.TargetDefinitionResolutionException: Could not find "org.eclipse.swtbot.feature.group/2.3.0.201506081302" in the repositories of the current location"

Am I missing something?

**Comment by apupier** [comment]
this is strange, the version mentioned is available here: http://download.eclipse.org/technology/swtbot/releases/2.3.0/features/?d

**Comment by rherrmann** [comment]
> this is strange, the version mentioned is available here: http://download.eclipse.org/technology/swtbot/releases/2.3.0/features/?d

But the target definition file references http://download.eclipse.org/technology/swtbot/releases/latest/ (because I couldn't find another repository). If there are per-version repos, we should probably specify on of those (v2.3?).

**Comment by apupier** [comment]
the latest release is http://download.eclipse.org/technology/swtbot/releases/2.4.0/

I'm not sure which version should be used for each of the Eclipse version.
We might deduced it from https://wiki.eclipse.org/SWTBot/Releases

**Comment by rherrmann** [comment]
Hm, I can't get the SWTBot tests to work at all.
I tried 4.3, 4.6 targets with SWTBot versions 2.2, 2.3 and 2.4 but the first test  (`PreferencesTest::should_update_test_package_prefix_when_preferences_change`) 'hangs'. The preference dialog opens, but the _MoreUnit > Java_ page that should be shown next gets not selected. The UI stays responsive.

**Comment by rherrmann** [comment]
I've opened an issue #25 to attach a pull request to. Currently, I am running the bot tests for each target platform and it looks like v2.4 works for all Eclipse versions. Up until now they are (mostly) green. I'll upload a pull request as soon as I'm done.

**Pull request merged**